### PR TITLE
add form field help option

### PR DESCRIPTION
**Pull request by gimler**
see https://github.com/symfony/symfony/pull/3147

### CODE REVIEW

File: Resources/config/field_type_help.xml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+        <parameter key="simple_things_form_extra.form.extension.help.class">SimpleThings\FormExtraBundle\Form\Extension\FieldTypeHelpExtension</parameter>
+    </parameters>
+
+    <services>
+        <service id="simple_things_form_extra.form.extension.help" class="%simple_things_form_extra.form.extension.help.class%">
+            <tag name="form.type_extension" alias="field" />
+        </service>
+
+    </services>
+</container>
```
Comment (stof) [inline Resources/config/field_type_help.xml:17]: why using a separate field if it is always loaded ?

File: Resources/views/Form/field_type_help.html.twig [hunk 2]
```diff
@@ -0,0 +1,65 @@
+{% use 'form_div_layout.html.twig' %}
```
Comment (stof) [inline Resources/views/Form/field_type_help.html.twig]: this should be removed as it breaks things as soon as you use another theme between them.
And it is not needed due to the way the form theming works

Comment (stof) [inline Resources/views/Form/field_type_help.html.twig]: and why not using the existing form theme provided by the bundle ?

Comment (gimler) [inline Resources/views/Form/field_type_help.html.twig]: Ok i right i have test this. So my problem is that i wan't call parent() but this works only when you extends or use another template "Calling "parent" on a template that does not extend nor "use" another template is forbidden"
Seperate file my intention was to keep the feature seperate.
But you are right the feature is always loaded so i change it and merge it into the  existing one.

Is there a way that my form theme extends a form theme e.q. MopaBootstrapBundle:Form:fields.html.twig and import the SimpleThingsFormExtraBundle:Form:field_type_help.html.twig but didn't override all the mopa code?

File: Resources/config/field_type_help.xml (hunk 1 above)
Comment (gimler) [inline Resources/config/field_type_help.xml:17]: i use this code as example https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/FrameworkBundle/Resources/config/form.xml#L138

Comment (gimler) [inline Resources/config/field_type_help.xml:17]: You need the alias without it it did not load the extension

Comment (stof) [inline Resources/config/field_type_help.xml:17]: sorry, I wrote it too fast. I meant `a separate file`

Comment (gimler) [inline Resources/config/field_type_help.xml:17]: ok i merge it into form_extra.xml

**Comment by beberlei** [comment]
Can you make the type optional by adding a paramter in the DIC config?

```
simple_things_form_extra:
  help_extension: true
```

The Extra template is needed then and should override/extend the "field_row" template, nothing else.

**Comment by gimler** [comment]
At the moment i test to set a twig var base_template and set default so when you use the template in your own you can define the base_template var.

I think override the field_row didn't work no access to help, but i will test that.

**Comment by gimler** [comment]
ok use cannot handle a var see https://github.com/fabpot/Twig/issues/500 but field_row works

File: Resources/views/Form/field_type_help.html.twig [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    {% endif %}
+{% endblock %}
+
+{% block field_row %}
+{% spaceless %}
+    <div>
+        {{ form_label(form, label|default(null)) }}
+        {{ form_errors(form) }}
+        {{ form_widget(form) }}
+        {{ form.vars.help }}
```
Comment (stof) [inline Resources/views/Form/field_type_help.html.twig]: `{{ help }}` would be enough

Comment (stof) [inline Resources/views/Form/field_type_help.html.twig]: oh, btw, it should be `{{ block('field_help') }}` otherwise the previous block is useless

Comment (gimler) [inline Resources/views/Form/field_type_help.html.twig]: fixed

File: DependencyInjection/SimpleThingsFormExtraExtension.php [hunk 4]
```diff
@@ -18,6 +18,10 @@ public function load(array $configs, ContainerBuilder $container)
         $processor = new Processor();
         $config = $processor->processConfiguration(new Configuration(), $configs);
 
+        if ($config['help_extension']) {
```
Comment (beberlei) [inline DependencyInjection/SimpleThingsFormExtraExtension.php:21]: You have to add this as booleanValue() in the Configuration.php

File: README.md [hunk 5]
```diff
@@ -197,6 +197,35 @@ $builder->add('body', 'textarea', array(
...
+
+``` yaml
+# app/config/config.yml
+simple_things_form_extra:
+    help_extension: true
+```
+
+
+```
+{% use 'SimpleThingsFormExtraBundle:Form:field_type_help.html.twig' %}
```
Comment (beberlei) [inline README.md:220]: You could also suggest to add this to config.yml for global help message:

```
twig:
   form.
     resources: ["SimpleThingsFormExtraBundle:Form:field_type_help.html.twig"]
```

File: DependencyInjection/SimpleThingsFormExtraExtension.php (hunk 4 above)
Comment (gimler) [inline DependencyInjection/SimpleThingsFormExtraExtension.php:21]: fixed

File: README.md (hunk 5 above)
Comment (gimler) [inline README.md:220]: fixed

**Comment by gimler** [comment]
ping

**Comment by beberlei** [comment]
sorry, forgot :-)

**Pull request merged**